### PR TITLE
Feat/liquidity manager

### PR DESCRIPTION
**Pull request by kinrezC**
- adds liquidity manager contract for zapping into positions

### CODE REVIEW

File: src/LiquidityManager.sol [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            ERC20(tokenIn).transferFrom(msg.sender, address(this), amountTokenIn);
+            amountSyOut += sy.deposit(receiver, tokenIn, amountTokenIn, 0);
+        }
+
+        if (amountSyOut < minSyMinted) {
+            revert InsufficientSYMinted(amountSyOut, minSyMinted);
+        }
+    }
+
+    function allocateFromSy(IRMM rmm, uint256 amountSy, uint256 minPtOut, uint256 minLiquidityDelta, uint256 initialGuess, uint256 epsilon) external returns (uint256 liquidity) {
```
Comment (Alexangelj) [inline src/LiquidityManager.sol:74]: Address arguments should be addresses, not interfaces. interfaces make it clunky

File: src/LiquidityManager.sol [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        ERC20 sy = ERC20(address(rmm.SY()));
+        ERC20 pt = ERC20(address(rmm.PT()));
+
+        PYIndex index = IPYieldToken(rmm.YT()).newIndex();
+        uint256 rX = rmm.reserveX();
+        uint256 rY = rmm.reserveY();
+
+        // validate swap approximation
+        (uint256 syToSwap,) = computeSyToPtToAddLiquidity(rmm, rX, rY, index, amountSy, block.timestamp, initialGuess, epsilon);
```
Comment (Alexangelj) [inline src/LiquidityManager.sol:83]: Why do we compute syToSwap?

File: src/LiquidityManager.sol (hunk 1 above)
Comment (Alexangelj) [reply src/LiquidityManager.sol:74]: Multiple places with this

File: src/LiquidityManager.sol [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+import { RMM, IPYieldToken } from "./RMM.sol";
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {IStandardizedYield} from "pendle/interfaces/IStandardizedYield.sol";
+import {PYIndexLib, PYIndex} from "pendle/core/StandardizedYield/PYIndex.sol";
+import {FixedPointMathLib} from "solmate/utils/FixedPointMathLib.sol";
+
+import "./lib/RmmErrors.sol";
+
+interface IRMM {
```
Comment (Alexangelj) [inline src/LiquidityManager.sol:12]: Since we already import RMM, we can cast it to RMM

**Review by Alexangelj — COMMENTED** [review]
The calls seem pretty straightforward:
- pull tokens from user
- approve rmm to spend those pulled tokens
- do rmm swap
- approve swap output tokens to be spent by rmm
- do allocate

Which gives me an idea:
- add multicall and make a public function for each single step (i.e. swap, allocate).
- in frontend, we build a multicall of: deposit tokens (pull from user), swap, allocate
- this would make it so the only difference between allocate from sy or allocate from pt is in the swap call we put in the multicall

Make sure there is no usage of msg.value in multicall stack trace

File: src/LiquidityManager.sol (hunk 3 above)
Comment (kinrezC) [reply src/LiquidityManager.sol:12]: That was causing weird stack issues with via-ir, no clue why

**Review by Alexangelj — COMMENTED** [review]
Missing reentrancy guard and the tests dont test anything

**Pull request merged**